Nettle 4: hash *_digest no longer takes an output-length argument.

--- a/common/rfb/CSecurityRSAAES.cxx	2026-08-17 07:12:37.814270830 +0200
+++ b/common/rfb/CSecurityRSAAES.cxx	2026-08-17 07:12:56.146880604 +0200
@@ -26,6 +26,7 @@
 #endif
 
 #include <stdlib.h>
+#include <string.h>
 #ifndef WIN32
 #include <unistd.h>
 #endif
@@ -227,7 +228,9 @@
   sha1_update(&ctx, 4, lenServerKey);
   sha1_update(&ctx, serverKey.size, serverKeyN);
   sha1_update(&ctx, serverKey.size, serverKeyE);
-  sha1_digest(&ctx, sizeof(f), f);
+  uint8_t _sha1full[SHA1_DIGEST_SIZE];
+  sha1_digest(&ctx, _sha1full);
+  memcpy(f, _sha1full, sizeof(f));
   const char *title = "Server key fingerprint";
   std::string text = core::format(
     "The server has provided the following identifying information:\n"
@@ -305,24 +308,27 @@
     sha1_init(&ctx);
     sha1_update(&ctx, 16, clientRandom);
     sha1_update(&ctx, 16, serverRandom);
-    sha1_digest(&ctx, 16, key);
+    uint8_t _sha1full[SHA1_DIGEST_SIZE];
+    sha1_digest(&ctx, _sha1full);
+    memcpy(key, _sha1full, 16);
     rais = new rdr::AESInStream(rawis, key, 128);
     sha1_init(&ctx);
     sha1_update(&ctx, 16, serverRandom);
     sha1_update(&ctx, 16, clientRandom);
-    sha1_digest(&ctx, 16, key);
+    sha1_digest(&ctx, _sha1full);
+    memcpy(key, _sha1full, 16);
     raos = new rdr::AESOutStream(rawos, key, 128);
   } else {
     struct sha256_ctx ctx;
     sha256_init(&ctx);
     sha256_update(&ctx, 32, clientRandom);
     sha256_update(&ctx, 32, serverRandom);
-    sha256_digest(&ctx, 32, key);
+    sha256_digest(&ctx, key);
     rais = new rdr::AESInStream(rawis, key, 256);
     sha256_init(&ctx);
     sha256_update(&ctx, 32, serverRandom);
     sha256_update(&ctx, 32, clientRandom);
-    sha256_digest(&ctx, 32, key);
+    sha256_digest(&ctx, key);
     raos = new rdr::AESOutStream(rawos, key, 256);
   }
   if (isAllEncrypted)
@@ -357,7 +363,7 @@
     sha1_update(&ctx, 4, lenServerKey);
     sha1_update(&ctx, serverKey.size, serverKeyN);
     sha1_update(&ctx, serverKey.size, serverKeyE);
-    sha1_digest(&ctx, hashSize, hash);
+    sha1_digest(&ctx, hash);
   } else {
     hashSize = 32;
     struct sha256_ctx ctx;
@@ -368,7 +374,7 @@
     sha256_update(&ctx, 4, lenServerKey);
     sha256_update(&ctx, serverKey.size, serverKeyN);
     sha256_update(&ctx, serverKey.size, serverKeyE);
-    sha256_digest(&ctx, hashSize, hash);
+    sha256_digest(&ctx, hash);
   }
   raos->writeBytes(hash, hashSize);
   raos->flush();
@@ -405,7 +411,7 @@
     sha1_update(&ctx, 4, lenClientKey);
     sha1_update(&ctx, clientKey.size, clientKeyN);
     sha1_update(&ctx, clientKey.size, clientKeyE);
-    sha1_digest(&ctx, hashSize, realHash);
+    sha1_digest(&ctx, realHash);
   } else {
     struct sha256_ctx ctx;
     sha256_init(&ctx);
@@ -415,7 +421,7 @@
     sha256_update(&ctx, 4, lenClientKey);
     sha256_update(&ctx, clientKey.size, clientKeyN);
     sha256_update(&ctx, clientKey.size, clientKeyE);
-    sha256_digest(&ctx, hashSize, realHash);
+    sha256_digest(&ctx, realHash);
   }
   if (memcmp(hash, realHash, hashSize) != 0)
     throw protocol_error("Hash doesn't match");
--- a/common/rfb/SSecurityRSAAES.cxx	2026-08-17 07:12:37.816112981 +0200
+++ b/common/rfb/SSecurityRSAAES.cxx	2026-08-17 07:12:56.146982965 +0200
@@ -27,6 +27,7 @@
 #include <errno.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <assert.h>
 
 #include <vector>
@@ -412,24 +413,27 @@
     sha1_init(&ctx);
     sha1_update(&ctx, 16, serverRandom);
     sha1_update(&ctx, 16, clientRandom);
-    sha1_digest(&ctx, 16, key);
+    uint8_t _sha1full[SHA1_DIGEST_SIZE];
+    sha1_digest(&ctx, _sha1full);
+    memcpy(key, _sha1full, 16);
     rais = new rdr::AESInStream(rawis, key, 128);
     sha1_init(&ctx);
     sha1_update(&ctx, 16, clientRandom);
     sha1_update(&ctx, 16, serverRandom);
-    sha1_digest(&ctx, 16, key);
+    sha1_digest(&ctx, _sha1full);
+    memcpy(key, _sha1full, 16);
     raos = new rdr::AESOutStream(rawos, key, 128);
   } else {
     struct sha256_ctx ctx;
     sha256_init(&ctx);
     sha256_update(&ctx, 32, serverRandom);
     sha256_update(&ctx, 32, clientRandom);
-    sha256_digest(&ctx, 32, key);
+    sha256_digest(&ctx, key);
     rais = new rdr::AESInStream(rawis, key, 256);
     sha256_init(&ctx);
     sha256_update(&ctx, 32, clientRandom);
     sha256_update(&ctx, 32, serverRandom);
-    sha256_digest(&ctx, 32, key);
+    sha256_digest(&ctx, key);
     raos = new rdr::AESOutStream(rawos, key, 256);
   }
   if (isAllEncrypted)
@@ -464,7 +468,7 @@
     sha1_update(&ctx, 4, lenClientKey);
     sha1_update(&ctx, clientKey.size, clientKeyN);
     sha1_update(&ctx, clientKey.size, clientKeyE);
-    sha1_digest(&ctx, hashSize, hash);
+    sha1_digest(&ctx, hash);
   } else {
     hashSize = 32;
     struct sha256_ctx ctx;
@@ -475,7 +479,7 @@
     sha256_update(&ctx, 4, lenClientKey);
     sha256_update(&ctx, clientKey.size, clientKeyN);
     sha256_update(&ctx, clientKey.size, clientKeyE);
-    sha256_digest(&ctx, hashSize, hash);
+    sha256_digest(&ctx, hash);
   }
   raos->writeBytes(hash, hashSize);
   raos->flush();
@@ -512,7 +516,7 @@
     sha1_update(&ctx, 4, lenServerKey);
     sha1_update(&ctx, serverKey.size, serverKeyN);
     sha1_update(&ctx, serverKey.size, serverKeyE);
-    sha1_digest(&ctx, hashSize, realHash);
+    sha1_digest(&ctx, realHash);
   } else {
     struct sha256_ctx ctx;
     sha256_init(&ctx);
@@ -522,7 +526,7 @@
     sha256_update(&ctx, 4, lenServerKey);
     sha256_update(&ctx, serverKey.size, serverKeyN);
     sha256_update(&ctx, serverKey.size, serverKeyE);
-    sha256_digest(&ctx, hashSize, realHash);
+    sha256_digest(&ctx, realHash);
   }
   if (memcmp(hash, realHash, hashSize) != 0)
     throw protocol_error("Hash doesn't match");
--- a/common/rfb/CSecurityDH.cxx	2026-08-17 07:12:37.817836521 +0200
+++ b/common/rfb/CSecurityDH.cxx	2026-08-17 07:12:37.834891884 +0200
@@ -125,7 +125,7 @@
   struct md5_ctx md5Ctx;
   md5_init(&md5Ctx);
   md5_update(&md5Ctx, sharedSecret.size(), sharedSecret.data());
-  md5_digest(&md5Ctx, 16, key);
+  md5_digest(&md5Ctx, key);
   struct aes128_ctx aesCtx;
   aes128_set_encrypt_key(&aesCtx, key);
 
